typegate/src/errors.ts

```diff
@@ -1,6 +1,141 @@
 // Copyright Metatype OÜ, licensed under the Elastic License 2.0.
 // SPDX-License-Identifier: Elastic-2.0
 
+import { basename, dirname } from "std/url/mod.ts";
+import { extname } from "std/path/mod.ts";
+import { getLogger } from "./log.ts";
+
+const logger = getLogger(import.meta);
+
+export enum ErrorKind {
+  User = "user", // from user input -- 4xx
+  Service = "service", // from foreign service -- 500, 503, 504
+  Typegraph = "typegraph", // bad typegraph, should be fixed by the typegraph writer -- 500
+  System = "system", // internal services
+  Typegate = "typegate", // unexpected error (bug?) -- 500
+}
+
+export class BaseError extends Error {
+  public module: string | null;
+  public code: number;
+  #type: string | null = null;
+
+  constructor(
+    module: ImportMeta | string | null,
+    public kind: ErrorKind,
+    message: string,
+    code: number | null = null,
+  ) {
+    super(message);
+
+    if (module != null && typeof module === "object") {
+      const bname = basename(module.url);
+      const dname = basename(dirname(module.url));
+      this.module = `${dname}/${bname.replace(extname(bname), "")}`;
+    } else {
+      this.module = module;
+    }
+
+    if (!code) {
+      switch (kind) {
+        case ErrorKind.User:
+          this.code = 400;
+          break;
+        default:
+          this.code = 500;
+      }
+    } else {
+      this.code = code;
+    }
+  }
+
+  withType(type: string): this {
+    this.#type = type;
+    return this;
+  }
+
+  toResponse(graphqlFormat = true): Response {
+    const type = this.#type ?? this.constructor.name;
+    logger.error(
```

Comment:
ideally, in non debug (config) we shall hide the details of the error to the user but that might also reduce the ux (for now at least)

Reply:
Yeah, the current approach is very...unpolished like that. Something consistent and more crafted is in order.